banlist updates

Diapolo opened the pull request:
banlist: update set dirty to be more fine grained
- move the SetBannedSetDirty(false) call from DumpData() into DumpBanlist()
- ensure we only set false, if the write succeeded

banlist: better handling of banlist in StartNode()
- only start working on/with banlist data, if reading in the banlist from
  disk didn't fail
- as CNode::setBannedIsDirty is false (default) when reading fails, we
  don't need to explicitly set it to false to prevent writing banlist.dat
  in that case either

banlist: add more banlist infos to log / add GUI signal
- to match the peers.dat handling also supply a debug.log entry for how
  many entries were loaded from banlist.dat and how long it took
- add a GUI init message for loading the banlist (same as with peers.dat)
- move the same message for peers.dat upwards in the code, to be able to
  reuse the timing variable nStart and also just log, if our read from
  peers.dat didn't fail

banlist (bugfix): allow CNode::SweepBanned() to run on interval
- allows CNode::SweepBanned() to run, even if !CNode::BannedSetIsDirty(),
  because if nBanUntil is over we want the ban to be disabled for these
  nodes

Commit:
banlist: update set dirty to be more fine grained

- move the SetBannedSetDirty(false) call from DumpData() into DumpBanlist()
- ensure we only set false, if the write succeeded

Commit:
banlist: better handling of banlist in StartNode()

- only start working on/with banlist data, if reading in the banlist from
  disk didn't fail
- as CNode::setBannedIsDirty is false (default) when reading fails, we
  don't need to explicitly set it to false to prevent writing
  banlist.dat in that case either

Commit:
banlist: add more banlist infos to log / add GUI signal

- to match the peers.dat handling also supply a debug.log entry for how
  many entries were loaded from banlist.dat and how long it took
- add a GUI init message for loading the banlist (same as with peers.dat)

- move the same message for peers.dat upwards in the code, to be able to
  reuse the timing variable nStart and also just log, if our read from
  peers.dat didn't fail

jonasschnelli commented on src/net.cpp:
```diff
@@ -1705,31 +1702,37 @@ void static Discover(boost::thread_group& threadGroup)
         LogPrintf("Invalid or missing banlist.dat; recreating\n");
+    } else {
+        CNode::SetBanned(banmap); // thread save setter
+        CNode::SetBannedSetDirty(false); // no need to write down, just read data
```
This line `CNode::SetBannedSetDirty(false);` should be called even if `bandb.Read(ban map)` failed.

Diapolo replied:
As written in the code, it is false as default, so it isn't dirty, because no functions were called that change the flag :).

jonasschnelli replied:
Why is it's default false (`net.cppL448)?
I would recommend to call `CNode::SetBannedSetDirty(false);` when we start up a node. There is no case where the list can be "dirty" in this state.

Diapolo replied:
IMHO an not explicitly initialized bool defaults to false (0).

jonasschnelli commented:
utACK (mind the nit above).

Diapolo commented on src/net.cpp:
```diff
@@ -1705,31 +1702,37 @@ void static Discover(boost::thread_group& threadGroup)
     }
 
-    //try to read stored banlist
+    uiInterface.InitMessage(_("Loading banned addresses..."));
```
Should this be ``ips/subnets`` rather than ``addresses``?

jonasschnelli replied:
Hmm.... it could be `_("Loading banned addresses/subnets...")` but if we be particular about this: a subnet defines an amount of "addresses". But i don't care about that.

laanwj replied:
`"Loading ban list..."`

Commit:
banlist (bugfix): allow CNode::SweepBanned() to run on interval

- allows CNode::SweepBanned() to run, even if !CNode::BannedSetIsDirty(),
  because if nBanUntil is over we want the ban to be disabled for these
  nodes

Diapolo commented:
@jonasschnelli Would be nice if you could test this pull and review commit 4, so this can get merged soon. I've got further small fixes for this but don't want to blow up this pull ;).

Diapolo commented:
@jonasschnelli Ping

Diapolo commented:
@jonasschnelli @laanwj Can we get this in please?

Diapolo commented:
@jonasschnelli Too sad you couldn't yet ACK this, mind taking a look?

jgarzik commented on src/net.cpp:
```diff
@@ -2470,28 +2473,31 @@ bool CBanDB::Read(banmap_t& banSet)
-             banmap.size(), GetTimeMillis() - nStart);
+        LogPrint("net", "Flushed %d banned node ips/subnets to banlist.dat  %dms\n",
+            banmap.size(), GetTimeMillis() - nStart);
+    }
```
General principle:  Code like "if (!dirty) return;" reduces the indentation level for the standard case, and makes the code more readable.

Diapolo replied:
So you would have written?
```
if (bandb.Write(banmap)) CNode::SetBannedSetDirty(false);
```

MarcoFalke replied:
I think it was `if (!CNode::BannedSetIsDirty()) return;`

Diapolo replied:
LOL, got it now ;)... thanks for that hint.

dcousens replied:
@Diapolo any chance you'll do the early exit as @jgarzik suggested?

jgarzik commented:
ut ACK

Diapolo commented:
@jgarzik Thanks for review :).

dcousens commented on src/net.cpp:
```diff
@@ -555,11 +555,13 @@ void CNode::SweepBanned()
     banmap_t::iterator it = setBanned.begin();
     while(it != setBanned.end())
     {
+        CSubNet subNet = (*it).first;
```
`const`?

Diapolo replied:
Not used a single time currently in net.cpp with (*it), so I'll leave that. But I'm open for further comments on this.

dcousens commented on src/net.cpp:
```diff
@@ -1901,31 +1898,37 @@ void static Discover(boost::thread_group& threadGroup)
     int64_t nStart = GetTimeMillis();
     {
         CAddrDB adb;
         if (!adb.Read(addrman))
```
nit: Avoid the negation in logic here now that we handle both cases?

Diapolo replied:
OK, will push a new version.

dcousens commented on src/net.cpp:
```diff
@@ -1901,31 +1898,37 @@ void static Discover(boost::thread_group& threadGroup)
     CBanDB bandb;
     banmap_t banmap;
-    if (!bandb.Read(banmap))
+    if (!bandb.Read(banmap)) {
```
nit: same here?

Diapolo replied:
Yeah, will also push a new version ;).

dcousens commented on src/net.cpp:
```diff
@@ -2470,28 +2473,31 @@ bool CBanDB::Read(banmap_t& banSet)
 void DumpBanlist()
 {
-    int64_t nStart = GetTimeMillis();
+    CNode::SweepBanned(); // clean unused entires (if bantime has expired)
```
`s/entires/entries`

dcousens commented:
utACK

Diapolo commented:
@dcousens @jgarzik Reworked to fix your valuable nits, can you re-check and re-ACK please?

dcousens commented:
@Diapolo re-reviewed commit-by-commit.  utACK / concept ACK.

Diapolo commented:
Seems unrelated!?

```
Running 161 test cases...

No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.

The build has been terminated
```

Diapolo commented:
What is holding up the merge here?
@jgarzik Ping :)

dcousens commented:
utACK

laanwj commented:
All utACK but I can't really see from the code changes whether it's correct. No one has tested this yet? Or is this covered by the current tests?

Diapolo commented:
Well, I tested it :-D and I also was aware of these things while @jonasschnelli implemented it, but AFAIK no one was interrested in fixing/improving at that time...

jonasschnelli commented:
utACK. A unittest would be nice (or extending the rpc test with some banlist.dat checking on file basis).
